### PR TITLE
Clean up and refactor search code.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
This PR reduces the formerly large `SearchFieldMixin` to include only what needs to be in a mixin (the state management for initializing and disposing search components). I moved controller functionality to the controller mixins `SearchControllerMixin` and `AutoCompleteSearchControllerMixin` and moved widget helper methods into formal Widgets `SearchField` and `AutoCompleteSearchField`.

RELEASE_NOTE_EXCEPTION=[not user facing]

### CODE REVIEW

File: packages/devtools_app/lib/src/shared/ui/search.dart [hunk 1]
```diff
@@ -62,6 +61,14 @@ mixin SearchControllerMixin<T extends DataSearchStateMixin> {
...
   Duration? get debounceDelay => null;
 
+  /// Text field controller for the [SearchField] that this instance of
+  /// [SearchControllerMixin] controls.
+  SearchTextEditingController? searchTextFieldController;
+
+  /// Focus node for the [SearchField] that this instance of
+  /// [SearchControllerMixin] controls.
+  FocusNode? searchFieldFocusNode;
+
```
Comment (elliette) [inline packages/devtools_app/lib/src/shared/ui/search.dart:71]: When would these be `null` (why are they nullable?)

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/shared/ui/search.dart:71]: These are nulled out in dispose and re-initialized in initSearch. The enclosing controller mixin will live longer than the search field for which the text field and the focus node are used. However, there are several places in the controller that need to access the text field controller and the focus node, so we store them in the controller, and then init and dispose them with the widget lifecycle (this is what SearchFieldMixin does). I don't love this solution, but it seemed cleaner than what we had.

Comment (elliette) [reply packages/devtools_app/lib/src/shared/ui/search.dart:71]: Got it, that makes sense. I'm mostly trying to figure out if there is a way having all the null-assert `!s` for these fields. Would it be possible to change these to be `late` fields (initialized in `initSearch`) and disposed of but not set to `null` during dispose? Or do they need to be reset back to `null`?

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/shared/ui/search.dart:71]: Setting to null ensures that we don't try to dispose an already disposed object, whereas using late wouldn't protect against that. For example, we dispose these objects in initSearch before re-initializing them, and we wouldn't be able to do that if we used late.

File: packages/devtools_app/lib/src/shared/ui/search.dart [hunk 2]
```diff
@@ -19,6 +19,8 @@ import '../theme.dart';
 import '../ui/utils.dart';
 import '../utils.dart';
 
+// TODO(kenz): break this file up into managable pieces.
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/ui/search.dart:22]: Should this have an issue associated with it?

File: packages/devtools_app/lib/src/shared/ui/search.dart (hunk 1 above)
Comment (elliette) [reply packages/devtools_app/lib/src/shared/ui/search.dart:71]: Got it, thanks! Optional - switch these (and `rawKeyboardFocusNode` below) to private nullable fields, and add corresponding non-nullable getter for them. Besides that, LGTM!

**Review by elliette — APPROVED** [review]
LGTM with one suggestion

File: packages/devtools_app/lib/src/shared/ui/search.dart [hunk 3]
```diff
@@ -673,6 +720,30 @@ mixin AutoCompleteSearchControllerMixin on SearchControllerMixin {
...
+
+  void selectFromSearchField(String selection) {
+    searchTextFieldController!.clear();
+    search = selection;
+    clearSearchField(force: true);
+    selectTheSearch = true;
+    closeAutoCompleteOverlay();
+  }
+
+  void clearSearchField({force = false}) {
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/ui/search.dart:732]: This parameter should have a type.

**Review by bkonyi — APPROVED** [review]

**Pull request merged**